bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/ThingManager.java

```diff
@@ -997,25 +987,60 @@ protected void activate(ComponentContext componentContext) {
 
     protected void addThingHandlerFactory(ThingHandlerFactory thingHandlerFactory) {
         logger.debug("Thing handler factory '{}' added", thingHandlerFactory.getClass().getSimpleName());
-
         thingHandlerFactories.add(thingHandlerFactory);
+        handleThingHandlerFactoryAddition(getBundleName(thingHandlerFactory));
+    }
 
-        for (Thing thing : things) {
-            if (thingHandlerFactory.supportsThingType(thing.getThingTypeUID())) {
-                if (!isHandlerRegistered(thing)) {
-                    registerAndInitializeHandler(thing, thingHandlerFactory);
-                } else {
-                    logger.debug("Thing handler for thing '{}' already registered", thing.getUID());
-                }
-            }
+    private Set<String> loadedXmlThingTypes = new CopyOnWriteArraySet<>();
+
+    protected void addReadyMarker(ReadyMarker readyMarker, Map<String, Object> properties) {
+        if (properties.containsKey(ReadyMarker.XML_THING_TYPE)) {
+            String bsn = (String) properties.get(ReadyMarker.XML_THING_TYPE);
+            loadedXmlThingTypes.add(bsn);
+            handleThingHandlerFactoryAddition(bsn);
+        }
+    }
+
+    protected void removeReadyMarker(ReadyMarker readyMarker, Map<String, Object> properties) {
+        if (properties.containsKey(ReadyMarker.XML_THING_TYPE)) {
+            String bsn = (String) properties.get(ReadyMarker.XML_THING_TYPE);
+            loadedXmlThingTypes.remove(bsn);
         }
     }
 
+    private void handleThingHandlerFactoryAddition(String bsn) {
+        thingHandlerFactories.stream().filter(it -> {
+            return getBundleName(it).equals(bsn);
+        }).forEach(thingHandlerFactory -> {
+            things.forEach(thing -> {
+                if (thingHandlerFactory.supportsThingType(thing.getThingTypeUID())) {
+                    if (!isHandlerRegistered(thing)) {
+                        registerAndInitializeHandler(thing, thingHandlerFactory);
+                    } else {
+                        logger.debug("Thing handler for thing '{}' already registered", thing.getUID());
+                    }
+                }
+            });
+        });
+    }
+
+    private String getBundleName(ThingHandlerFactory it) {
```

Comment:
Why does the argument use the name "it"? It doesn't matter, but IMHO uncommon to use "iterator" (??) for an variable name of an argument.